File: src/lib/routing.ts
import type { LatLng, RouteResult } from '../types'
import { haversineM } from './geo'

/**
 * FOSSGIS community OSRM instance — same API as the project demo server
 * (router.project-osrm.org) but with far better uptime. Overridable in Settings.
 */
export const DEFAULT_ROUTING_BASE_URL = 'https://routing.openstreetmap.de/routed-car'

const FALLBACK_AVG_KMH = 70
const FALLBACK_POINTS = 64

/** Straight-line route used when a routing provider is unreachable. */
export function greatCircleRoute(from: LatLng, to: LatLng, avgKmh = FALLBACK_AVG_KMH): RouteResult {
  const polyline: LatLng[] = []
  for (let i = 0; i <= FALLBACK_POINTS; i++) {
    const t = i / FALLBACK_POINTS
    polyline.push({
      lat: from.lat + (to.lat - from.lat) * t,
      lng: from.lng + (to.lng - from.lng) * t,
    })
  }
  const distanceM = haversineM(from, to)
  return {
    polyline,
    distanceM,
    durationS: (distanceM / 1000 / avgKmh) * 3600,
    approximate: true,
  }
}

/** Query an OSRM server. Throws on any non-ok response or missing route. */
export async function osrmRoute(
  from: LatLng,
  to: LatLng,
  baseUrl: string,
  fetchFn: typeof fetch = fetch,
): Promise<RouteResult> {
  const coords = `${from.lng},${from.lat};${to.lng},${to.lat}`
  const url = `${baseUrl.replace(/\/$/, '')}/route/v1/driving/${coords}?overview=full&geometries=geojson`
  const res = await fetchFn(url)
  if (!res.ok) throw new Error(`routing HTTP ${res.status}`)
  const data = (await res.json()) as {
    routes?: Array<{ distance: number; duration: number; geometry: { coordinates: [number, number][] } }>
  }
  const route = data.routes?.[0]
  if (!route) throw new Error('routing: no route in response')
  return {
    polyline: route.geometry.coordinates.map(([lng, lat]) => ({ lat, lng })),
    distanceM: route.distance,
    durationS: route.duration,
    approximate: false,
  }
}

/** Get a route, degrading to a straight-line path if the provider fails. */
export async function getRoute(
  from: LatLng,
  to: LatLng,
  baseUrl: string,
  fetchFn: typeof fetch = fetch,
): Promise<RouteResult> {
  try {
    return await osrmRoute(from, to, baseUrl, fetchFn)
  } catch {
    return greatCircleRoute(from, to)
  }
}

File: cli/routing.py
"""Routing with a great-circle fallback — mirrors src/lib/routing.ts."""

from __future__ import annotations

from dataclasses import dataclass

import requests

from .geo import Point, haversine_m

DEFAULT_ROUTING_BASE_URL = "https://routing.openstreetmap.de/routed-car"
FALLBACK_AVG_KMH = 70
FALLBACK_POINTS = 64
REQUEST_TIMEOUT = 12


@dataclass
class RouteResult:
    polyline: list[Point]
    distance_m: float
    duration_s: float
    approximate: bool


def great_circle_route(a: Point, b: Point, avg_kmh: float = FALLBACK_AVG_KMH) -> RouteResult:
    polyline = [
        (a[0] + (b[0] - a[0]) * (i / FALLBACK_POINTS), a[1] + (b[1] - a[1]) * (i / FALLBACK_POINTS))
        for i in range(FALLBACK_POINTS + 1)
    ]
    distance_m = haversine_m(a, b)
    return RouteResult(polyline, distance_m, (distance_m / 1000 / avg_kmh) * 3600, True)


def osrm_route(a: Point, b: Point, base_url: str) -> RouteResult:
    coords = f"{a[1]},{a[0]};{b[1]},{b[0]}"
    url = f"{base_url.rstrip('/')}/route/v1/driving/{coords}?overview=full&geometries=geojson"
    resp = requests.get(url, timeout=REQUEST_TIMEOUT)
    resp.raise_for_status()
    routes = resp.json().get("routes") or []
    if not routes:
        raise ValueError("routing: no route in response")
    route = routes[0]
    polyline = [(lat, lng) for lng, lat in route["geometry"]["coordinates"]]
    return RouteResult(polyline, route["distance"], route["duration"], False)


def get_route(a: Point, b: Point, base_url: str = DEFAULT_ROUTING_BASE_URL) -> RouteResult:
    try:
        return osrm_route(a, b, base_url)
    except (requests.RequestException, ValueError, KeyError):
        return great_circle_route(a, b)
